### DISCUSSION TITLE
Wox v2 plans

### DISCUSSION THREAD

**Discussion by qianlifeng** [Wox-launcher/Wox#3937] [General]
# Why v2?
When Wox was first created, it was mainly on Windows, and after few years, I switched to macos since 2015, and  the development of Wox has been handed over to Wox launcher organization, mainly @bao-qian. Thanks for his contributions to Wox over the years. 

Wox has been neglected in recent years, the last release was back in 2020. It's time to take Wox to the next level, a cross-platform Wox v2. 

# Changes
- I've learned a few things from the contributions of Wox, one in particular is not to put resources in a non-Github location, as they will become invalid and outdated over time as they migrate. All future Wox resources will be hosted on Github, including the plugin store, theme store, documentation, discussions and more!

- Wox v1 was developed mainly in dotnet, because Wox was mainly used on windows. Although now dotnet can also be cross-platform, but obviously it is not the optimal framework. I want the core of Wox to be separate from the UI presentation. The core should be stable (compared to the cross-platform shift of dotnet), able to support the next 10 years of development. We chose Golang as the core framwork for Wox v2, and most of the functionality will be in this core. 

  UI technology is changing rapidly, we should not rely too much on the UI framework, we chose to use ~~Tauri + React~~ Flutter as the UI interface now, if there is a newer UI technology (tauri, swiftui, maui, etc.), we can switch over without too much cost.

# Version compatibility
Because the development framework is different and v2 is cross platform, so v2 and v1 (plugin) will have some compatibility issues, theoretically this compatibility problem can be solved by providing a "dotnet legency host" in v2. We will try to provide such a compatibility layer, to ensure that most of the v1 plugins can still work on windows. However, it is important to note that this is not a 100% compatibility issue, and some features will still need to be adapted by developers.

# v1 bug fix
We now have no resources to fix v1 bugs, so any exceptions reported will not been fixed.

# Dev plans
It was planed to spend 3-4 month to develop Wox v2. ~~It will be great if we can bring v2 GA before the end of the year~~. 

Now we (@lylwx, @qianlifeng) have released the first tech preview for mac(windows and linux version will be provided later), if you are using macos, please [download](https://github.com/Wox-launcher/Wox/releases) and try it out, and give us feedback, thanks!

**Comment by jhdxr** [comment]
Nice to see it has been picked up. I've been using v1 for the past years. It's not perfect but it's still the best on Windows IMO. (Yes I have tried other alternatives like utools). 

I'm a bit interested in the plan for plugins. Python support on v1 seems to be a failure. As a developer mainly working on Windows, I don't really install any dev envs on my host machine any more, it has been isolated/installed in Docker, WSL or remote dev machines. I think nowdays it's not fesiable to ask users to do extra setup to run certain plugins. Even dev tools like jetbrains are bundling with JDK instead of asking users to setup. Not to mention the version management, which is another chaos.

**Reply by qianlifeng** [reply]
In v1, c# plugin is the main plugin development language, and python can only be used as a second-class citizen, can not provide the same development experience and functionality with c# (such as plugin settings, query efficiency).

In V2, all plugin development languages will be first class citizens, we will provide a variety of different plugin Host ([nodejs](https://github.com/Wox-launcher/Wox/tree/v2/Wox.Plugin.Host.Nodejs), python, c#, java and more), which will make them all have the same plugin development experience and functionality.

Of course, as you said, this brings up the issue of language SDK installation dependencies. Some of the [core plugins](https://github.com/Wox-launcher/Wox/tree/v2/Wox/plugin/system) in v2 will be developed in golang and bundled with the Wox distribution. So, even if users don't install any dependencies, they can still get a basic experience. If a user wants to use a plugin written in another language, he has to install the SDK for that language. I don't want to spend too much effort bundling these SDKs into Wox, as it would become complicated (jetbrains only need to bundle JDK, but we need to bundle nodejs, python, dotnet, jdk...). We can guide the user to download and install these SDK dependencies, and installing these SDKs shouldn't be very complicated nowadays, just next, next, done.

**Reply by eugenesvk** [reply]
by the way, have you considered a single WASM host for the "non-native" plugins? Would still allow writing plugins in many languages, but wouldn't depend on language-specific SDKs to use

**Reply by qianlifeng** [reply]
Yes, I considered using WASM as a unified plugin host, but the current development of the WASM runtime is still in flux, and [WASI](https://github.com/WebAssembly/WASI/blob/main/Proposals.md) hasn't been finalized yet, things may change. so at this stage I don't want to use WASM as the host. If WASM develops well in the future, we can add WASM host alongside other hosts, but not replace them.

**Reply by SrGeneroso** [reply]
Oh boy, I downloaded the v2 on windows and is was ok, so I went to install the v1. @jhdxr is right, installing python is not right.

**Reply by eissar** [reply]
> Some of the [core plugins](https://github.com/Wox-launcher/Wox/tree/v2/Wox/plugin/system) in v2 will be developed in golang and bundled with the Wox distribution.

very excited with the core plugins already written. looking much better than the powertoys-run plugin system which requires so much object-orientated boilerplate.

how will golang plugins be packaged? will they compile to library .so files in the .wox directory? or just executables?

**Comment by korenevskiy** [comment]
Please add search service "[Yandex](www.ya.ru)" in v2

**Reply by qianlifeng** [reply]
Do you mean websearch using Yandex? pls open another issue so we can discuss there more.

**Reply by korenevskiy** [reply]
yes

**Comment by M142857** [comment]
Happy, congratulations and look forward to it!
开心 祝贺 期待

**Comment by f1h0** [comment]
You will give competition to the powertoys program.

**Comment by WangJincheng4869** [comment]
哎~有点晚了啊，其他竞品现在做的已经非常好了，我现在在用uTools，希望wox2能让我回心转意

**Reply by qianlifeng** [reply]
> 哎~有点晚了啊，其他竞品现在做的已经非常好了，我现在在用uTools，希望wox2能让我回心转意

晚不晚的无所谓，哪怕只有一个人用。只是不忍心这个项目就这样停止了。Wox也不存在跟谁竞争的问题，大家喜欢用哪个都可以，只要有帮助，用得顺手，那所有的这些开源软件都值得

**Reply by Kikkwa** [reply]
uTools 太垃圾了，没多少人用的，只要 Wox 做的流畅一些，省资源很容易就把 uTools 替代了

**Reply by Cateds** [reply]
utools主要是太丑了加上绑定一些服务，感觉有点别扭
还是期待wox吧，好饭不怕晚

**Reply by HsuDan** [reply]
还是期待wox，utools我用过了，一大堆没用的服务，我只希望能象wox这样用最简单的方式直接达成我的目的，而不是在utools里那样还要先点一下面板选项，然后再输，搞那么麻烦我还不如不用呢。所以还是希望wox能发展起来！加油！另，要是wox2能象wox1那样直观就好了，或者有个使用手册也行啊，讲实话wox2的设置实在是有点看不明白，比如在哪里设置默认的浏览器，在哪里连接everything服务，我点来点去搞了二十多分钟还是没明白，最终又用回了wox1

**Reply by JensenPeng** [reply]
> 还是期待wox，utools我用过了，一大堆没用的服务，我只希望能象wox这样用最简单的方式直接达成我的目的，而不是在utools里那样还要先点一下面板选项，然后再输，搞那么麻烦我还不如不用呢。所以还是希望wox能发展起来！加油！另，要是wox2能象wox1那样直观就好了，或者有个使用手册也行啊，讲实话wox2的设置实在是有点看不明白，比如在哪里设置默认的浏览器，在哪里连接everything服务，我点来点去搞了二十多分钟还是没明白，最终又用回了wox1

everything现在还不支持。所以我已经切回了wox1

**Reply by qianlifeng** [reply]
实际在最新的编译版本中 https://github.com/Wox-launcher/Wox/actions/runs/14293759445 files插件(搜索`f `)就是直接调用的everything搜索文件的

**Comment by archon810** [comment]
Given that Microsoft based its PowerToys Run utility on Wox, 1) Do you expect them to switch to the new implementation as well? and 2) Is there a commitment from MS to help development of Wox?

**Reply by qianlifeng** [reply]
1. No, since Wox is using the MIT open source license, it's up to them to decide which implementation to use. IMO, it is not easy for them to change the implementation easily as they are a company, they have more responsibility and stabilisation requirement.
2. No

**Reply by lukaszjablonski** [reply]
Did for PowerToys Run Microsoft change files and folder search engine to native Windows Search? I don't think they use Everything Search for that, right?

**Comment by cantonadong** [comment]
mac 上已经有 alfred ，而且非常非常非常非常非常非常完善和好用了，为什么要维护？
win上几乎没有体验和 alfred 水平的软件，除了 wox，为什么不把精力放在 win上？

**Reply by qianlifeng** [reply]
因为我主用mac, 我太不可能去维护一个我自己不用的软件. 这也是为什么当初我把Wox交给社区去维护的原因

**Comment by SullieLore** [comment]
It's exciting to see Wox v2. I have always kept an eye out for news about Wox, as it always seemed to run the best. Looking forward to trying out the Windows version when it's released. Congratulations to the team for coming back and I hope for all the best with your work.

**Comment by hoiwanchang** [comment]
活久见的激动人心的消息。虽然我从win迁移到了mac，现在又从mac迁移到了linux，但希望WOX能够越做越好！
真心希望后续能够支持wasm，能够为插件做贡献的人也会更多
还有，有锈化的计划么？

**Reply by qianlifeng** [reply]
多谢支持，等版本稳定下来后会考虑wasm的

**Comment by lvxingqi** [comment]
windows版本的v2版本现在好像用不了，再等等

**Comment by Karlies** [comment]
> 你已经激活了Wox 4226 次
- 打开Wox的关于，我看到了这个。
- 我是从大概15年开始使用Wox的，那时的我希望拥有一个Alfred的Windows替代品，然后我找到了Wox。检索、翻译、剪贴板、终端的功能给我的工作生活提供了极大的便利。
- 在Wox没有更新的那段时间，不少插件的功能出现了些问题，类似集成有道的词典也不顶用了。我也尝试过uTools，但是不能所见即所得，每次都要进入二级窗口让我不到1个小时就还是回归了Wox。
- 可惜的是我没有开发其中的一些插件，也就去Theme文件夹做了几套自用的主题，但是好像也没有很好的分享的渠道。
- 能在年前看到Wox 2.0的消息真是太好了，希望后续能试用上and尝试结合文档开发插件。

感谢作者以及Wox这么多年的陪伴！

**Comment by johnchenkzy** [comment]
wox v2 UI now is use flutter or use Tauri + React ? From source code i see both of them.

**Reply by qianlifeng** [reply]
We have tried flutter, Tauri, Electron, and we choose flutter as the final UI implementation.

**Comment by johnchenkzy** [comment]
Is there any place to discuss the implementation of the Windows version of Wox v2? I found that in the v2 version, Wox parses the .lnk files to find the specific paths and then indexes them. However, the implementation using golnk has some issues with handling Unicode and some lnk files cannot be correctly parsed (If lnk file has no linkInfo structure,but has LinkTargetIDList structure). Is it really necessary to parse lnk files, or can we directly call lnk? If parsing lnk files is indeed necessary, I can pull some PR to upstream golnk.

**Reply by johnchenkzy** [reply]
And If I want to contribution for wox v2,Which things do I need to pay attention to?

**Reply by qianlifeng** [reply]
You are right, it is not necessary to parse the lnk file. Maybe We could index lnk file itself if we could get lnk ico someway. Windows version of Wox is not as great as Macos version now. There are two main function we need to resolve now: 
1. [App parse not implement](https://github.com/Wox-launcher/Wox/blob/3c91038c5b2428f8b1019f0a3aa660d9572bc14f/Wox/plugin/system/app/app_windows.go#L39)
2. [Transparent effect not works well ](https://github.com/Wox-launcher/Wox/blob/3c91038c5b2428f8b1019f0a3aa660d9572bc14f/Wox.UI.Flutter/wox/lib/main.dart#L77)

PR is welcome😀, thx!

**Comment by CescMessi** [comment]
本来想找文档写个调用大模型的插件，没想到发现了新版的消息，而且插件里已经有了ChatGPT相关的代码，期待早日用上正式版~

**Comment by JohnWilliston** [comment]
Oh wow, I am so happy to learn of this new project! Thanks to all who are contributing!

**Comment by ZJunSong** [comment]
Thanks for this awesome project, I have used wox for 3 yrs. Hope to see Wox v2 in windows asap.

**Comment by edanastas** [comment]
I still see a lot of features lacking from Launchbar on a mac. which is pretty much the gold standard. Wish we can get something more like Launchbar on windows.

**Comment by FuRuF-11** [comment]
wox v1可能是我最喜欢的windows软件了，感谢作者多年来的努力！
